docs/api/console.md

```diff
@@ -0,0 +1,159 @@
+
+Each of these has a similar interface. Each function accepts an arbitrary number of arguments. The logged message consists of a human readable representation of the arguments, separated by strings.
+
+The following two lines will both log the message "Hello World".
+```typescript
+console.log("Hello World!");
+console.log("Hello", "World!");
+```
+
+In practice using multiple arguments is unwieldy and difficult to read.
+Most javascript guildelines advise using string interpolation instead of multiple arguments.
+
+For example the following
+``` typescript
+function sayHello(name: string) {
+  console.log(`Hello ${name} from jstz!`);
+}
+```
+should be preferred to
+
+``` typescript
+function sayHello(name: string) {
+  console.log("Hello", name, "from jstz!");
+}
+```
+
+Different logging functions can be combined to make logs more informative and to allow easy filtering.
+
+```typescript
+function myAwesomeFunction(arg: unknown) {
+    console.debug(`Entered myAwesomeFunction with argument ${arg}`)
+    if (typeof arg !== 'string') {
+    console.warn("Converting non string argument to string");
+    arg = arg.toString();
+    }
+    try {
+        let result = functionThatMayThrow(arg);
+        console.info(`Result: ${result}`);
+    } catch (error) {
+        console.error(`Call failed with message ${error}`);
+        throw error;
+    }
+}
+```
+### Assertions
+
+`console.assert` will log an error message if its first argument is falsy.
+It will not throw an error or interrupt control flow, but can be used to flag places where 
+something unexpected occurs.
+If the first argument is truthy there will be no effect.
+
+```typescript
+function divide(a : number, b : number) : number {
+   console.assert(b != 0, "Trying to divide by 0");
+   return a / b;
+}
+```
+
+
+### Groups
+
+In a more complex program we may wish to create logs in a structured way.
+
+`console.group()` increases the level of indentation of the logs, along with a group name.
+The group may be closed using `console.groupEnd()`.
+Internally the console maintains a stack of group names, the level of indentation is simply the current stack height.
+This can be used to organise the logs as desired.
+For example the following program performs a brute force search for Pythagorean triples, reflecting the level of nesting using groups.
+```typescript
+function pythagoreanTriples (limit = 100) {
+  for (let x = 2; x <= limit; ++x) {
+    try {
+      console.group(`x = ${x}`);
+      for (let y = 1; y < x; ++y) {
+        try {
+          console.group(`y = ${y}`);
+          for (let z = x; z < x + y; ++z) {
+            console.log(`trying z = ${z}`)
+            if (x*x + y*y === z*z) {
+              console.log(`Success! ${x}^2 + ${y}^2 = ${z}^2`)
+              break;
+            } 
+          }
+        } finally {
+          console.groupEnd();
+        }
+      }
+    } finally {
+      console.groupEnd();
+    }
+  }
+}
+```
```

Comment:
While `try`-`finally` is likely the most robust way to ensure groups are closed, I think it might make the example clearer and easier to parse (especially when I don't see any operations that could raise an exception)